src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -563,5 +621,44 @@ void RayletBasedActorScheduler::HandleWorkerLeaseReply(
   }
 }
 
+void GcsActorScheduler::HandleWorkerLeaseRejectedReply(
+    std::shared_ptr<GcsActor> actor, const rpc::RequestWorkerLeaseReply &reply) {
+  // The request was rejected because of insufficient resources.
+  auto node_id = actor->GetNodeID();
+  auto &cluster_resource_manager =
+      cluster_task_manager_->GetClusterResourceScheduler()->GetClusterResourceManager();
+  auto acquired_resources = actor->GetMutableAcquiredResources();
+  cluster_resource_manager.AddNodeAvailableResources(scheduling::NodeID(node_id.Binary()),
+                                                     *acquired_resources);
+  acquired_resources->Clear();
+  actor->UpdateAddress(rpc::Address());
+  if (normal_task_resources_changed_callback_) {
+    normal_task_resources_changed_callback_(node_id, reply.resources_data());
+  }
```

Comment:
This only works for gcs based scheduling. With your previous change, now raylet based scheduling can also have rejected reply for spilled lease request. We need to handle them separately.